docs/styleguides/python.md

```diff
@@ -0,0 +1,259 @@
+# DC/OS Python Style Guide
+###### Version 0.1
+###### April-6-2017
+
+For Python source code within DC/OS project, we prefer to keep [Google Python Style Guide Revision 2.59] as the base and
+**extend** it with some changes that will make DC/OS codebase look modern, consistent and easy to follow.
+
+##### Rationale
+
+With any style guide, the primary requirement is to adopt one and follow it consistently. With our choices below, we
+can use tools available in the python ecosystem to our advantage, enforce consistency in our codebase, while creating
+and publishing automatically generated documentation from the code itself.
+
+##### Versioned document
+
+Styleguide _conventions_ are not set in stone. We can improve upon this or remove inconvenient styles by creating a PR,
+discussing the merits or demerits to arrive at a consensus. Developers who contribute code to the DC/OS repo are the
+main stake holders of this style guide.
+
+##### Python Version Requirement
+
+This style guide is written for Python3.5 and above. As new features of the language get standardized, this style guide
+can be revised to adopt new features.
+
+#### Base Style Guide
+
+* [Google Python Style Guide Revision 2.59]
+
+
+#### DC/OS Specific extensions to Base Style Guide
+
+* Import Requirements
+    * Adopt the base style guide preferences.
+
+        ```
+        Use import x for importing packages and modules.
+        Use from x import y where x is the package prefix and y is the module name with no prefix.
+        Use from x import y as z if two modules named y are to be imported or if y is an inconveniently long name.
+        ```
+    * Use `__init__.py` only to represent python packages.
+        * Prefer empty `__init__.py` modules. Keep definitions in submodules.
+
+    _Rationale_: Keeping `__init__.py` empty helps in avoid unnecessary imports when you import a submodule from
+    package. Plus,it helps in navigating the code base you know that an object is strictly defined in a submodule and
+    not aliased in `__init__.py`
+
+* Import formatting
+
+    * The types of imports should be separated by single line and order should be.
+
+        1. standard library imports.
+        2. third-party imports.
+        3. application specific imports.
+
+    * The `from` style import and `import` statements should be separated by single line too.
```

Comment:
I will remove this line and the requirement. I agree with some of the points made here. (Let's stick to the essential).

Also, there is enough negative votes to this ( @mellenburg , @vespian  and @branden ).

My guess is, I was having this bias from days when `isort` tool was not available and I had to take help of IDE/ editor settings to do automated sorting.

I ran this small experiment.

* Start with unsorted imports
```
import test_util.cluster
import traceback
from pkgpanda.util import logger, write_string
from retrying import retry, RetryError
from subprocess import CalledProcessError

import json
import logging
import os
from teamcity.messages import TeamcityServiceMessages
from test_util.cluster import run_bootstrap_nginx
from test_util.dcos_api_session import DcosApiSession, DcosUser
from test_util.helpers import CI_CREDENTIALS, marathon_app_id_to_mesos_dns_subdomain, random_id
from typing import Callable, List
import pprint
import random
import retrying

import sys
import test_util.aws
import uuid
```
* sort using isort with no separation between types.

```isort -o retrying fixed_with_groups_no_lines.py```

Output

```
import json
import logging
import os
import pprint
import random
import sys
import traceback
import uuid
from subprocess import CalledProcessError
from typing import Callable, List

import retrying
from retrying import RetryError, retry

import test_util.aws
import test_util.cluster
from pkgpanda.util import logger, write_string
from teamcity.messages import TeamcityServiceMessages
from test_util.cluster import run_bootstrap_nginx
from test_util.dcos_api_session import DcosApiSession, DcosUser
from test_util.helpers import (CI_CREDENTIALS,
                               marathon_app_id_to_mesos_dns_subdomain,
                               random_id)
```

* Sort using isort with group and line separation

```
isort -o retrying -lbt 1 fixed_with_group_and_lines.py
```

Output.

```
import json
import logging
import os
import pprint
import random
import sys
import traceback
import uuid

from subprocess import CalledProcessError
from typing import Callable, List

import retrying

from retrying import RetryError, retry

import test_util.aws
import test_util.cluster

from pkgpanda.util import logger, write_string
from teamcity.messages import TeamcityServiceMessages
from test_util.cluster import run_bootstrap_nginx
from test_util.dcos_api_session import DcosApiSession, DcosUser
from test_util.helpers import (CI_CREDENTIALS,
                               marathon_app_id_to_mesos_dns_subdomain,
                               random_id)
```

* It seems. it just fine to adopt the 2nd style - that is one without any separation between `import` and `from` types.